client/network/src/protocol/sync.rs

```diff
@@ -895,6 +974,47 @@ impl<B: BlockT> ChainSync<B> {
 				}
 			}
 		}
+		if let Some(sync) = &self.warp_sync {
+			if sync.is_complete() {
+				return None
+			}
+			if let Some(request) = sync.next_state_request() {
+				for (id, peer) in self.peers.iter_mut() {
+					if peer.state.is_available() && peer.best_number >= sync.target_block_number() {
+						trace!(target: "sync", "New StateRequest for {}", id);
+						peer.state = PeerSyncState::DownloadingState;
+						return Some((id.clone(), request))
+					}
+				}
+			}
+		}
+		None
+	}
+
+	/// Get a warp sync request, if any.
+	pub fn warp_sync_request(&mut self) -> Option<(PeerId, WarpProofRequest<B>)> {
+		if self
+			.peers
+			.iter()
+			.any(|(_, peer)| peer.state == PeerSyncState::DownloadingWarpProof)
+		{
+			// Only one pending state request is allowed.
+			return None
+		}
+		if let Some(sync) = &self.warp_sync {
+			if sync.is_complete() {
+				return None
+			}
+			if let Some(request) = sync.next_warp_poof_request() {
+				for (id, peer) in self.peers.iter_mut() {
+					if peer.state.is_available() && peer.best_number >= sync.target_block_number() {
```

Comment:
`target_block_number` always return `0` until we hit the phase of downloading the state proof. So, we could also send such a warp proof request to some peer that itself is still syncing?

Sounds like a bug?

Reply:
Yest. Unfortunately with the current GRANDPA warp sync protocol it is not possible to learn the latest finalized block number before downloading and verifying all proofs first. In my first implementation I've set a recently announced block as target, but in general that would not work because there's no way to know if it is finalized in advance. I'll add a generic check that a peer is not too far behind the majority of the peerset instead.